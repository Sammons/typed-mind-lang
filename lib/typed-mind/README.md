# @sammons/typed-mind

Core language parser and validator for TypedMind DSL - a declarative language for building software architectures.

## Installation

```bash
npm install @sammons/typed-mind
```

## Usage

```typescript
import { TypedMindParser, TypedMindValidator } from '@sammons/typed-mind';

// Parse TypedMind source code
const parser = new TypedMindParser();
const program = parser.parse(source);

// Validate the parsed program
const validator = new TypedMindValidator();
const validationResult = validator.validate(program);

if (validationResult.isValid) {
  console.log('Program is valid!');
} else {
  console.log('Validation errors:', validationResult.errors);
}
```

## What is TypedMind?

TypedMind is a Domain Specific Language (DSL) for declaratively describing software architectures, including:

- **Data Transfer Objects (DTOs)** - Define data structures and their validation rules
- **Service classes** - Specify business logic and their dependencies
- **UI components** - Describe user interface elements and their relationships
- **Asset management** - Define static assets and routing configuration
- **Cross-cutting concerns** - Handle validation, security, and other system-wide concerns

## Features

- **Declarative syntax** - Focus on what your system does, not how it's implemented
- **Strong validation** - Catch architectural issues before they become problems
- **Dependency tracking** - Understand relationships between components
- **Import system** - Modular architecture with file-based organization
- **Type safety** - Comprehensive validation of all language constructs

## Example

```typedmind
program "User Management System"

dto UserDto {
  id: string
  email: string
  name: string
  createdAt: date
}

service UserService {
  method createUser(email: string, name: string): UserDto
  method getUserById(id: string): UserDto
  depends on DatabaseService
}

ui UserListComponent {
  displays UserDto[]
  triggers UserService.createUser
}
```

## Grammar

The complete TypedMind grammar is documented in [grammar.md](./grammar.md) and available in EBNF format in [grammar.ebnf](./grammar.ebnf).

## Requirements

- Node.js >= 22.0.0

## Related Packages

- [@sammons/typed-mind-cli](https://www.npmjs.com/package/@sammons/typed-mind-cli) - Command-line interface
- [@sammons/typed-mind-lsp](https://www.npmjs.com/package/@sammons/typed-mind-lsp) - Language Server Protocol implementation
- [@sammons/typed-mind-renderer](https://www.npmjs.com/package/@sammons/typed-mind-renderer) - HTML/D3.js renderer

## License

MIT