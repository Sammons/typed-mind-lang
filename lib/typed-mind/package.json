{
  "name": "@sammons/typed-mind",
  "version": "1.0.0",
  "description": "Core language parser and validator for TypedMind DSL",
  "author": "sammons",
  "license": "MIT",
  "main": "dist/index.js",
  "types": "dist/index.d.ts",
  "files": [
    "dist"
  ],
  "scripts": {
    "build": "tsc --build",
    "dev": "tsc --build --watch",
    "clean": "rm -rf dist *.tsbuildinfo",
    "test": "vitest --run",
    "test:watch": "vitest --watch",
    "lint": "eslint src"
  },
  "devDependencies": {
    "@types/node": "^22.10.5",
    "tsup": "^8.3.5",
    "vitest": "^2.1.8"
  },
  "publishConfig": {
    "access": "public"
  }
}