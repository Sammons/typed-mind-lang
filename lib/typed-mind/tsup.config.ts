import { defineConfig } from 'tsup';

export default defineConfig({
  entry: ['src/index.ts'],
  format: ['cjs'],
  dts: {
    resolve: true,
  },
  clean: true,
  sourcemap: true,
  splitting: false,
  treeshake: true,
  tsconfig: './tsconfig.json',
});