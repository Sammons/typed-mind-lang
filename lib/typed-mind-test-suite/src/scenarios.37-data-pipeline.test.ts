import { describe, it, expect } from 'vitest';
import { readFileSync } from 'fs';
import { join, dirname } from 'path';
import { fileURLToPath } from 'url';
import { DSLChecker } from '@sammons/typed-mind';

const __filename = fileURLToPath(import.meta.url);
const __dirname = dirname(__filename);

describe('scenario-37-data-pipeline', () => {
  const checker = new DSLChecker();
  const scenarioFile = 'scenario-37-data-pipeline.tmd';

  it('should validate data pipeline ETL architecture', () => {
    const content = readFileSync(join(__dirname, '..', 'scenarios', scenarioFile), 'utf-8');
    const result = checker.check(content);
    
    // Create a clean output for snapshots
    const output = {
      file: scenarioFile,
      valid: result.valid,
      errors: result.errors.map(err => ({
        line: err.position.line,
        column: err.position.column,
        message: err.message,
        severity: err.severity,
        suggestion: err.suggestion
      }))
    };
    
    expect(output).toMatchSnapshot();
  });
});