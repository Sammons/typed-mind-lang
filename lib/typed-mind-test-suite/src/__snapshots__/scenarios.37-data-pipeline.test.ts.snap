// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`scenario-37-data-pipeline > should validate data pipeline ETL architecture 1`] = `
{
  "errors": [
    {
      "column": 1,
      "line": 147,
      "message": "Orphaned entity 'SCHEMA_REGISTRY_URL'",
      "severity": "error",
      "suggestion": "Remove or reference this entity",
    },
    {
      "column": 1,
      "line": 148,
      "message": "Orphaned entity 'MONITORING_ENDPOINT'",
      "severity": "error",
      "suggestion": "Remove or reference this entity",
    },
    {
      "column": 1,
      "line": 161,
      "message": "Orphaned entity 'PYTHON_VERSION'",
      "severity": "error",
      "suggestion": "Remove or reference this entity",
    },
    {
      "column": 1,
      "line": 163,
      "message": "Orphaned entity 'JAVA_VERSION'",
      "severity": "error",
      "suggestion": "Remove or reference this entity",
    },
    {
      "column": 1,
      "line": 173,
      "message": "Orphaned entity 'ERROR_TOLERANCE'",
      "severity": "error",
      "suggestion": "Remove or reference this entity",
    },
    {
      "column": 1,
      "line": 175,
      "message": "Orphaned entity 'RETENTION_DAYS'",
      "severity": "error",
      "suggestion": "Remove or reference this entity",
    },
    {
      "column": 1,
      "line": 390,
      "message": "Orphaned entity 'TaskConfig'",
      "severity": "error",
      "suggestion": "Remove or reference this entity",
    },
    {
      "column": 1,
      "line": 397,
      "message": "Orphaned entity 'Dependency'",
      "severity": "error",
      "suggestion": "Remove or reference this entity",
    },
    {
      "column": 1,
      "line": 415,
      "message": "Orphaned entity 'ExecutionMetrics'",
      "severity": "error",
      "suggestion": "Remove or reference this entity",
    },
    {
      "column": 1,
      "line": 432,
      "message": "Orphaned entity 'DatasetMetadata'",
      "severity": "error",
      "suggestion": "Remove or reference this entity",
    },
    {
      "column": 1,
      "line": 438,
      "message": "Orphaned entity 'Schema'",
      "severity": "error",
      "suggestion": "Remove or reference this entity",
    },
    {
      "column": 1,
      "line": 443,
      "message": "Orphaned entity 'Field'",
      "severity": "error",
      "suggestion": "Remove or reference this entity",
    },
    {
      "column": 1,
      "line": 461,
      "message": "Orphaned entity 'CleaningRules'",
      "severity": "error",
      "suggestion": "Remove or reference this entity",
    },
    {
      "column": 1,
      "line": 471,
      "message": "Orphaned entity 'ValidationError'",
      "severity": "error",
      "suggestion": "Remove or reference this entity",
    },
    {
      "column": 1,
      "line": 476,
      "message": "Orphaned entity 'ValidationWarning'",
      "severity": "error",
      "suggestion": "Remove or reference this entity",
    },
    {
      "column": 1,
      "line": 480,
      "message": "Orphaned entity 'AggregateMetric'",
      "severity": "error",
      "suggestion": "Remove or reference this entity",
    },
    {
      "column": 1,
      "line": 485,
      "message": "Orphaned entity 'EnrichmentConfig'",
      "severity": "error",
      "suggestion": "Remove or reference this entity",
    },
    {
      "column": 1,
      "line": 489,
      "message": "Orphaned entity 'LookupConfig'",
      "severity": "error",
      "suggestion": "Remove or reference this entity",
    },
    {
      "column": 1,
      "line": 494,
      "message": "Orphaned entity 'APIEnrichment'",
      "severity": "error",
      "suggestion": "Remove or reference this entity",
    },
    {
      "column": 1,
      "line": 503,
      "message": "Orphaned entity 'ProcessingStats'",
      "severity": "error",
      "suggestion": "Remove or reference this entity",
    },
    {
      "column": 1,
      "line": 519,
      "message": "Orphaned entity 'FieldStatistics'",
      "severity": "error",
      "suggestion": "Remove or reference this entity",
    },
    {
      "column": 1,
      "line": 527,
      "message": "Orphaned entity 'DataPattern'",
      "severity": "error",
      "suggestion": "Remove or reference this entity",
    },
    {
      "column": 1,
      "line": 532,
      "message": "Orphaned entity 'QualityScore'",
      "severity": "error",
      "suggestion": "Remove or reference this entity",
    },
    {
      "column": 1,
      "line": 538,
      "message": "Orphaned entity 'AnomalyModel'",
      "severity": "error",
      "suggestion": "Remove or reference this entity",
    },
    {
      "column": 1,
      "line": 547,
      "message": "Orphaned entity 'Anomaly'",
      "severity": "error",
      "suggestion": "Remove or reference this entity",
    },
    {
      "column": 1,
      "line": 552,
      "message": "Orphaned entity 'AnomalySummary'",
      "severity": "error",
      "suggestion": "Remove or reference this entity",
    },
    {
      "column": 1,
      "line": 9,
      "message": "Import 'TaskRegistry' not found",
      "severity": "error",
      "suggestion": undefined,
    },
    {
      "column": 1,
      "line": 9,
      "message": "Import 'DependencyGraph' not found",
      "severity": "error",
      "suggestion": "Did you mean 'Dependency'?",
    },
    {
      "column": 1,
      "line": 13,
      "message": "Import 'CronScheduler' not found",
      "severity": "error",
      "suggestion": "Did you mean 'TaskScheduler'?",
    },
    {
      "column": 1,
      "line": 13,
      "message": "Import 'EventScheduler' not found",
      "severity": "error",
      "suggestion": "Did you mean 'TaskScheduler'?",
    },
    {
      "column": 1,
      "line": 17,
      "message": "Import 'StateManager' not found",
      "severity": "error",
      "suggestion": undefined,
    },
    {
      "column": 1,
      "line": 17,
      "message": "Import 'RetryPolicy' not found",
      "severity": "error",
      "suggestion": undefined,
    },
    {
      "column": 1,
      "line": 26,
      "message": "Import 'KafkaConsumer' not found",
      "severity": "error",
      "suggestion": undefined,
    },
    {
      "column": 1,
      "line": 26,
      "message": "Import 'SchemaRegistry' not found",
      "severity": "error",
      "suggestion": "Did you mean 'SCHEMA_REGISTRY_URL'?",
    },
    {
      "column": 1,
      "line": 30,
      "message": "Import 'S3Client' not found",
      "severity": "error",
      "suggestion": undefined,
    },
    {
      "column": 1,
      "line": 30,
      "message": "Import 'FileParser' not found",
      "severity": "error",
      "suggestion": undefined,
    },
    {
      "column": 1,
      "line": 34,
      "message": "Import 'HTTPClient' not found",
      "severity": "error",
      "suggestion": undefined,
    },
    {
      "column": 1,
      "line": 34,
      "message": "Import 'RateLimiter' not found",
      "severity": "error",
      "suggestion": undefined,
    },
    {
      "column": 1,
      "line": 38,
      "message": "Import 'DBConnector' not found",
      "severity": "error",
      "suggestion": undefined,
    },
    {
      "column": 1,
      "line": 38,
      "message": "Import 'QueryBuilder' not found",
      "severity": "error",
      "suggestion": undefined,
    },
    {
      "column": 1,
      "line": 47,
      "message": "Import 'ValidationRules' not found",
      "severity": "error",
      "suggestion": "Did you mean 'ValidationResult'?",
    },
    {
      "column": 1,
      "line": 47,
      "message": "Import 'CleansingStrategies' not found",
      "severity": "error",
      "suggestion": undefined,
    },
    {
      "column": 1,
      "line": 51,
      "message": "Import 'SchemaDefinitions' not found",
      "severity": "error",
      "suggestion": undefined,
    },
    {
      "column": 1,
      "line": 51,
      "message": "Import 'ValidationEngine' not found",
      "severity": "error",
      "suggestion": "Did you mean 'ValidationWarning'?",
    },
    {
      "column": 1,
      "line": 55,
      "message": "Import 'AggregationFunctions' not found",
      "severity": "error",
      "suggestion": undefined,
    },
    {
      "column": 1,
      "line": 55,
      "message": "Import 'WindowFunctions' not found",
      "severity": "error",
      "suggestion": undefined,
    },
    {
      "column": 1,
      "line": 59,
      "message": "Import 'LookupTables' not found",
      "severity": "error",
      "suggestion": undefined,
    },
    {
      "column": 1,
      "line": 59,
      "message": "Import 'ExternalAPIs' not found",
      "severity": "error",
      "suggestion": undefined,
    },
    {
      "column": 1,
      "line": 68,
      "message": "Import 'SparkSession' not found",
      "severity": "error",
      "suggestion": "Did you mean 'SPARK_VERSION'?",
    },
    {
      "column": 1,
      "line": 68,
      "message": "Import 'DataFrameOps' not found",
      "severity": "error",
      "suggestion": undefined,
    },
    {
      "column": 1,
      "line": 72,
      "message": "Import 'FlinkEnvironment' not found",
      "severity": "error",
      "suggestion": undefined,
    },
    {
      "column": 1,
      "line": 72,
      "message": "Import 'StreamOperators' not found",
      "severity": "error",
      "suggestion": undefined,
    },
    {
      "column": 1,
      "line": 76,
      "message": "Import 'BeamPipeline' not found",
      "severity": "error",
      "suggestion": "Did you mean 'createPipeline'?",
    },
    {
      "column": 1,
      "line": 76,
      "message": "Import 'Transforms' not found",
      "severity": "error",
      "suggestion": "Did you mean 'TransformersFile'?",
    },
    {
      "column": 1,
      "line": 85,
      "message": "Import 'StatisticalAnalyzer' not found",
      "severity": "error",
      "suggestion": undefined,
    },
    {
      "column": 1,
      "line": 85,
      "message": "Import 'PatternDetector' not found",
      "severity": "error",
      "suggestion": undefined,
    },
    {
      "column": 1,
      "line": 89,
      "message": "Import 'MLModels' not found",
      "severity": "error",
      "suggestion": undefined,
    },
    {
      "column": 1,
      "line": 89,
      "message": "Import 'ThresholdRules' not found",
      "severity": "error",
      "suggestion": undefined,
    },
    {
      "column": 1,
      "line": 93,
      "message": "Import 'MetricsCollector' not found",
      "severity": "error",
      "suggestion": "Did you mean 'TraceCollector'?",
    },
    {
      "column": 1,
      "line": 93,
      "message": "Import 'AlertManager' not found",
      "severity": "error",
      "suggestion": undefined,
    },
    {
      "column": 1,
      "line": 102,
      "message": "Import 'ParquetWriter' not found",
      "severity": "error",
      "suggestion": "Did you mean 'CacheWriter'?",
    },
    {
      "column": 1,
      "line": 102,
      "message": "Import 'DeltaWriter' not found",
      "severity": "error",
      "suggestion": undefined,
    },
    {
      "column": 1,
      "line": 106,
      "message": "Import 'RedshiftClient' not found",
      "severity": "error",
      "suggestion": "Did you mean 'REDSHIFT_CLUSTER'?",
    },
    {
      "column": 1,
      "line": 106,
      "message": "Import 'BigQueryClient' not found",
      "severity": "error",
      "suggestion": undefined,
    },
    {
      "column": 1,
      "line": 106,
      "message": "Import 'SnowflakeClient' not found",
      "severity": "error",
      "suggestion": undefined,
    },
    {
      "column": 1,
      "line": 110,
      "message": "Import 'RedisClient' not found",
      "severity": "error",
      "suggestion": undefined,
    },
    {
      "column": 1,
      "line": 110,
      "message": "Import 'MemcachedClient' not found",
      "severity": "error",
      "suggestion": undefined,
    },
    {
      "column": 1,
      "line": 114,
      "message": "Import 'KafkaProducer' not found",
      "severity": "error",
      "suggestion": undefined,
    },
    {
      "column": 1,
      "line": 114,
      "message": "Import 'KinesisClient' not found",
      "severity": "error",
      "suggestion": undefined,
    },
    {
      "column": 1,
      "line": 123,
      "message": "Import 'PrometheusClient' not found",
      "severity": "error",
      "suggestion": undefined,
    },
    {
      "column": 1,
      "line": 123,
      "message": "Import 'DatadogClient' not found",
      "severity": "error",
      "suggestion": undefined,
    },
    {
      "column": 1,
      "line": 127,
      "message": "Import 'FluentdClient' not found",
      "severity": "error",
      "suggestion": undefined,
    },
    {
      "column": 1,
      "line": 127,
      "message": "Import 'LogstashClient' not found",
      "severity": "error",
      "suggestion": undefined,
    },
    {
      "column": 1,
      "line": 131,
      "message": "Import 'JaegerClient' not found",
      "severity": "error",
      "suggestion": undefined,
    },
    {
      "column": 1,
      "line": 131,
      "message": "Import 'ZipkinClient' not found",
      "severity": "error",
      "suggestion": undefined,
    },
    {
      "column": 1,
      "line": 5,
      "message": "Entity 'runPipeline' is exported by multiple files: OrchestratorFile, BeamProcessorFile",
      "severity": "error",
      "suggestion": "Each entity should be exported by exactly one file. Remove the duplicate exports.",
    },
    {
      "column": 1,
      "line": 22,
      "message": "Entity 'KafkaIngestor' is exported by multiple files: IngestorsFile, KafkaIngestorFile",
      "severity": "error",
      "suggestion": "Each entity should be exported by exactly one file. Remove the duplicate exports.",
    },
    {
      "column": 1,
      "line": 22,
      "message": "Entity 'S3Ingestor' is exported by multiple files: IngestorsFile, S3IngestorFile",
      "severity": "error",
      "suggestion": "Each entity should be exported by exactly one file. Remove the duplicate exports.",
    },
    {
      "column": 1,
      "line": 22,
      "message": "Entity 'APIIngestor' is exported by multiple files: IngestorsFile, APIIngestorFile",
      "severity": "error",
      "suggestion": "Each entity should be exported by exactly one file. Remove the duplicate exports.",
    },
    {
      "column": 1,
      "line": 22,
      "message": "Entity 'DatabaseIngestor' is exported by multiple files: IngestorsFile, DatabaseIngestorFile",
      "severity": "error",
      "suggestion": "Each entity should be exported by exactly one file. Remove the duplicate exports.",
    },
    {
      "column": 1,
      "line": 43,
      "message": "Entity 'DataCleaner' is exported by multiple files: TransformersFile, DataCleanerFile",
      "severity": "error",
      "suggestion": "Each entity should be exported by exactly one file. Remove the duplicate exports.",
    },
    {
      "column": 1,
      "line": 43,
      "message": "Entity 'SchemaValidator' is exported by multiple files: TransformersFile, SchemaValidatorFile",
      "severity": "error",
      "suggestion": "Each entity should be exported by exactly one file. Remove the duplicate exports.",
    },
    {
      "column": 1,
      "line": 43,
      "message": "Entity 'Aggregator' is exported by multiple files: TransformersFile, AggregatorFile",
      "severity": "error",
      "suggestion": "Each entity should be exported by exactly one file. Remove the duplicate exports.",
    },
    {
      "column": 1,
      "line": 43,
      "message": "Entity 'Enricher' is exported by multiple files: TransformersFile, EnricherFile",
      "severity": "error",
      "suggestion": "Each entity should be exported by exactly one file. Remove the duplicate exports.",
    },
    {
      "column": 1,
      "line": 64,
      "message": "Entity 'SparkProcessor' is exported by multiple files: ProcessingFile, SparkProcessorFile",
      "severity": "error",
      "suggestion": "Each entity should be exported by exactly one file. Remove the duplicate exports.",
    },
    {
      "column": 1,
      "line": 64,
      "message": "Entity 'FlinkProcessor' is exported by multiple files: ProcessingFile, FlinkProcessorFile",
      "severity": "error",
      "suggestion": "Each entity should be exported by exactly one file. Remove the duplicate exports.",
    },
    {
      "column": 1,
      "line": 64,
      "message": "Entity 'BeamProcessor' is exported by multiple files: ProcessingFile, BeamProcessorFile",
      "severity": "error",
      "suggestion": "Each entity should be exported by exactly one file. Remove the duplicate exports.",
    },
    {
      "column": 1,
      "line": 81,
      "message": "Entity 'DataProfiler' is exported by multiple files: QualityFile, DataProfilerFile",
      "severity": "error",
      "suggestion": "Each entity should be exported by exactly one file. Remove the duplicate exports.",
    },
    {
      "column": 1,
      "line": 81,
      "message": "Entity 'AnomalyDetector' is exported by multiple files: QualityFile, AnomalyDetectorFile",
      "severity": "error",
      "suggestion": "Each entity should be exported by exactly one file. Remove the duplicate exports.",
    },
    {
      "column": 1,
      "line": 81,
      "message": "Entity 'QualityMonitor' is exported by multiple files: QualityFile, QualityMonitorFile",
      "severity": "error",
      "suggestion": "Each entity should be exported by exactly one file. Remove the duplicate exports.",
    },
    {
      "column": 1,
      "line": 98,
      "message": "Entity 'DataLakeWriter' is exported by multiple files: StorageFile, DataLakeWriterFile",
      "severity": "error",
      "suggestion": "Each entity should be exported by exactly one file. Remove the duplicate exports.",
    },
    {
      "column": 1,
      "line": 98,
      "message": "Entity 'WarehouseWriter' is exported by multiple files: StorageFile, WarehouseWriterFile",
      "severity": "error",
      "suggestion": "Each entity should be exported by exactly one file. Remove the duplicate exports.",
    },
    {
      "column": 1,
      "line": 98,
      "message": "Entity 'CacheWriter' is exported by multiple files: StorageFile, CacheWriterFile",
      "severity": "error",
      "suggestion": "Each entity should be exported by exactly one file. Remove the duplicate exports.",
    },
    {
      "column": 1,
      "line": 98,
      "message": "Entity 'StreamWriter' is exported by multiple files: StorageFile, StreamWriterFile",
      "severity": "error",
      "suggestion": "Each entity should be exported by exactly one file. Remove the duplicate exports.",
    },
    {
      "column": 1,
      "line": 119,
      "message": "Entity 'MetricsExporter' is exported by multiple files: MonitoringFile, MetricsExporterFile",
      "severity": "error",
      "suggestion": "Each entity should be exported by exactly one file. Remove the duplicate exports.",
    },
    {
      "column": 1,
      "line": 119,
      "message": "Entity 'LogAggregator' is exported by multiple files: MonitoringFile, LogAggregatorFile",
      "severity": "error",
      "suggestion": "Each entity should be exported by exactly one file. Remove the duplicate exports.",
    },
    {
      "column": 1,
      "line": 119,
      "message": "Entity 'TraceCollector' is exported by multiple files: MonitoringFile, TraceCollectorFile",
      "severity": "error",
      "suggestion": "Each entity should be exported by exactly one file. Remove the duplicate exports.",
    },
    {
      "column": 1,
      "line": 296,
      "message": "Call to 'KafkaConsumer.poll' references unknown entity 'KafkaConsumer'",
      "severity": "error",
      "suggestion": undefined,
    },
    {
      "column": 1,
      "line": 296,
      "message": "Call to 'SchemaRegistry.decode' references unknown entity 'SchemaRegistry'",
      "severity": "error",
      "suggestion": undefined,
    },
    {
      "column": 1,
      "line": 301,
      "message": "Call to 'S3Client.listObjects' references unknown entity 'S3Client'",
      "severity": "error",
      "suggestion": undefined,
    },
    {
      "column": 1,
      "line": 301,
      "message": "Call to 'FileParser.parse' references unknown entity 'FileParser'",
      "severity": "error",
      "suggestion": undefined,
    },
    {
      "column": 1,
      "line": 306,
      "message": "Call to 'HTTPClient.get' references unknown entity 'HTTPClient'",
      "severity": "error",
      "suggestion": undefined,
    },
    {
      "column": 1,
      "line": 306,
      "message": "Call to 'RateLimiter.checkLimit' references unknown entity 'RateLimiter'",
      "severity": "error",
      "suggestion": undefined,
    },
    {
      "column": 1,
      "line": 311,
      "message": "Call to 'DBConnector.connect' references unknown entity 'DBConnector'",
      "severity": "error",
      "suggestion": undefined,
    },
    {
      "column": 1,
      "line": 311,
      "message": "Call to 'QueryBuilder.build' references unknown entity 'QueryBuilder'",
      "severity": "error",
      "suggestion": undefined,
    },
    {
      "column": 1,
      "line": 316,
      "message": "Call to 'ValidationRules.apply' references unknown entity 'ValidationRules'",
      "severity": "error",
      "suggestion": undefined,
    },
    {
      "column": 1,
      "line": 316,
      "message": "Call to 'CleansingStrategies.execute' references unknown entity 'CleansingStrategies'",
      "severity": "error",
      "suggestion": undefined,
    },
    {
      "column": 1,
      "line": 321,
      "message": "Call to 'SchemaDefinitions.load' references unknown entity 'SchemaDefinitions'",
      "severity": "error",
      "suggestion": undefined,
    },
    {
      "column": 1,
      "line": 321,
      "message": "Call to 'ValidationEngine.validate' references unknown entity 'ValidationEngine'",
      "severity": "error",
      "suggestion": undefined,
    },
    {
      "column": 1,
      "line": 326,
      "message": "Call to 'AggregationFunctions.apply' references unknown entity 'AggregationFunctions'",
      "severity": "error",
      "suggestion": undefined,
    },
    {
      "column": 1,
      "line": 326,
      "message": "Call to 'WindowFunctions.compute' references unknown entity 'WindowFunctions'",
      "severity": "error",
      "suggestion": undefined,
    },
    {
      "column": 1,
      "line": 331,
      "message": "Call to 'LookupTables.join' references unknown entity 'LookupTables'",
      "severity": "error",
      "suggestion": undefined,
    },
    {
      "column": 1,
      "line": 331,
      "message": "Call to 'ExternalAPIs.fetch' references unknown entity 'ExternalAPIs'",
      "severity": "error",
      "suggestion": undefined,
    },
    {
      "column": 1,
      "line": 337,
      "message": "Call to 'SparkSession.createDataFrame' references unknown entity 'SparkSession'",
      "severity": "error",
      "suggestion": undefined,
    },
    {
      "column": 1,
      "line": 337,
      "message": "Call to 'DataFrameOps.transform' references unknown entity 'DataFrameOps'",
      "severity": "error",
      "suggestion": undefined,
    },
    {
      "column": 1,
      "line": 343,
      "message": "Call to 'FlinkEnvironment.addSource' references unknown entity 'FlinkEnvironment'",
      "severity": "error",
      "suggestion": undefined,
    },
    {
      "column": 1,
      "line": 343,
      "message": "Call to 'StreamOperators.apply' references unknown entity 'StreamOperators'",
      "severity": "error",
      "suggestion": undefined,
    },
    {
      "column": 1,
      "line": 349,
      "message": "Call to 'StatisticalAnalyzer.analyze' references unknown entity 'StatisticalAnalyzer'",
      "severity": "error",
      "suggestion": undefined,
    },
    {
      "column": 1,
      "line": 349,
      "message": "Call to 'PatternDetector.detect' references unknown entity 'PatternDetector'",
      "severity": "error",
      "suggestion": undefined,
    },
    {
      "column": 1,
      "line": 354,
      "message": "Call to 'MLModels.predict' references unknown entity 'MLModels'",
      "severity": "error",
      "suggestion": undefined,
    },
    {
      "column": 1,
      "line": 354,
      "message": "Call to 'ThresholdRules.check' references unknown entity 'ThresholdRules'",
      "severity": "error",
      "suggestion": undefined,
    },
    {
      "column": 1,
      "line": 359,
      "message": "Call to 'MetricsCollector.collect' references unknown entity 'MetricsCollector'",
      "severity": "error",
      "suggestion": undefined,
    },
    {
      "column": 1,
      "line": 359,
      "message": "Call to 'AlertManager.checkThresholds' references unknown entity 'AlertManager'",
      "severity": "error",
      "suggestion": undefined,
    },
    {
      "column": 1,
      "line": 364,
      "message": "Call to 'ParquetWriter.write' references unknown entity 'ParquetWriter'",
      "severity": "error",
      "suggestion": undefined,
    },
    {
      "column": 1,
      "line": 364,
      "message": "Call to 'DeltaWriter.merge' references unknown entity 'DeltaWriter'",
      "severity": "error",
      "suggestion": undefined,
    },
    {
      "column": 1,
      "line": 369,
      "message": "Call to 'RedshiftClient.copy' references unknown entity 'RedshiftClient'",
      "severity": "error",
      "suggestion": undefined,
    },
    {
      "column": 1,
      "line": 369,
      "message": "Call to 'BigQueryClient.load' references unknown entity 'BigQueryClient'",
      "severity": "error",
      "suggestion": undefined,
    },
    {
      "column": 1,
      "line": 374,
      "message": "Call to 'RedisClient.set' references unknown entity 'RedisClient'",
      "severity": "error",
      "suggestion": undefined,
    },
    {
      "column": 1,
      "line": 374,
      "message": "Call to 'MemcachedClient.set' references unknown entity 'MemcachedClient'",
      "severity": "error",
      "suggestion": undefined,
    },
    {
      "column": 1,
      "line": 377,
      "message": "Call to 'KafkaProducer.send' references unknown entity 'KafkaProducer'",
      "severity": "error",
      "suggestion": undefined,
    },
    {
      "column": 1,
      "line": 377,
      "message": "Call to 'KinesisClient.putRecords' references unknown entity 'KinesisClient'",
      "severity": "error",
      "suggestion": undefined,
    },
    {
      "column": 1,
      "line": 9,
      "message": "Export 'buildDAG' is not defined anywhere in the codebase",
      "severity": "error",
      "suggestion": "Define 'buildDAG' as a Function, Class, Constants, Asset, or UIComponent entity",
    },
    {
      "column": 1,
      "line": 9,
      "message": "Export 'validateDAG' is not defined anywhere in the codebase",
      "severity": "error",
      "suggestion": "Define 'validateDAG' as a Function, Class, Constants, Asset, or UIComponent entity",
    },
    {
      "column": 1,
      "line": 13,
      "message": "Export 'scheduleTask' is not defined anywhere in the codebase",
      "severity": "error",
      "suggestion": "Define 'scheduleTask' as a Function, Class, Constants, Asset, or UIComponent entity",
    },
    {
      "column": 1,
      "line": 13,
      "message": "Export 'cancelSchedule' is not defined anywhere in the codebase",
      "severity": "error",
      "suggestion": "Define 'cancelSchedule' as a Function, Class, Constants, Asset, or UIComponent entity",
    },
    {
      "column": 1,
      "line": 17,
      "message": "Export 'executeWorkflow' is not defined anywhere in the codebase",
      "severity": "error",
      "suggestion": "Define 'executeWorkflow' as a Function, Class, Constants, Asset, or UIComponent entity",
    },
    {
      "column": 1,
      "line": 123,
      "message": "Export 'exportMetrics' is not defined anywhere in the codebase",
      "severity": "error",
      "suggestion": "Define 'exportMetrics' as a Function, Class, Constants, Asset, or UIComponent entity",
    },
    {
      "column": 1,
      "line": 127,
      "message": "Export 'aggregateLogs' is not defined anywhere in the codebase",
      "severity": "error",
      "suggestion": "Define 'aggregateLogs' as a Function, Class, Constants, Asset, or UIComponent entity",
    },
    {
      "column": 1,
      "line": 131,
      "message": "Export 'collectTraces' is not defined anywhere in the codebase",
      "severity": "error",
      "suggestion": "Define 'collectTraces' as a Function, Class, Constants, Asset, or UIComponent entity",
    },
    {
      "column": 1,
      "line": 136,
      "message": "Export 'ConnectionConfig' is not defined anywhere in the codebase",
      "severity": "error",
      "suggestion": "Define 'ConnectionConfig' as a Function, Class, Constants, Asset, or UIComponent entity",
    },
    {
      "column": 1,
      "line": 136,
      "message": "Export 'SecurityConfig' is not defined anywhere in the codebase",
      "severity": "error",
      "suggestion": "Define 'SecurityConfig' as a Function, Class, Constants, Asset, or UIComponent entity",
    },
    {
      "column": 1,
      "line": 140,
      "message": "Export 'Serializer' is not defined anywhere in the codebase",
      "severity": "error",
      "suggestion": "Define 'Serializer' as a Function, Class, Constants, Asset, or UIComponent entity",
    },
    {
      "column": 1,
      "line": 140,
      "message": "Export 'Compressor' is not defined anywhere in the codebase",
      "severity": "error",
      "suggestion": "Define 'Compressor' as a Function, Class, Constants, Asset, or UIComponent entity",
    },
    {
      "column": 1,
      "line": 140,
      "message": "Export 'Encryptor' is not defined anywhere in the codebase",
      "severity": "error",
      "suggestion": "Define 'Encryptor' as a Function, Class, Constants, Asset, or UIComponent entity",
    },
    {
      "column": 1,
      "line": 140,
      "message": "Export 'RetryHandler' is not defined anywhere in the codebase",
      "severity": "error",
      "suggestion": "Define 'RetryHandler' as a Function, Class, Constants, Asset, or UIComponent entity",
    },
  ],
  "file": "scenario-37-data-pipeline.tmd",
  "valid": false,
}
`;
